### PR TITLE
Limit request size

### PR DESCRIPTION
**Pull request by schomatis**

### CODE REVIEW

File: handler.go [hunk 1]
```diff
@@ -111,7 +111,32 @@ func (s *RPCServer) handleReader(ctx context.Context, r io.Reader, w io.Writer,
 	}
 
 	var req request
-	if err := json.NewDecoder(r).Decode(&req); err != nil {
+	// Limit request size. Ideally this limit should be specific for each field
+	// in the JSON request but as simple defensive measure we just limit the
+	// entire HTTP body.
+	MAX_REQUEST_SIZE := 10 << 10 // FIXME: Agree on a value and extract.
```
Comment (magik6k) [inline handler.go:117]: Some Worker calls to things which compute snarks send ~20-30MB, so I'd probably set this to 100M

File: handler.go [hunk 2]
```diff
@@ -111,7 +111,32 @@ func (s *RPCServer) handleReader(ctx context.Context, r io.Reader, w io.Writer,
...
+	reqSize, err := buf.ReadFrom(io.LimitReader(r, int64(MAX_REQUEST_SIZE + 1)))
+	if err != nil {
+		// ReadFrom will ignore an EOF from LimitReader so this is an unexpected
+		// error.
+		rpcError(wf, &req, rpcParseError, xerrors.Errorf("reading request: %w", err))
+		return
+	}
+	if reqSize > int64(MAX_REQUEST_SIZE) {
+		// FIXME: Do we want to define a new error or can we consider reading errors
+		//  as part of the parsing stage (`rpcParseError`) and discriminate by error string?
```
Comment (magik6k) [inline handler.go:134]: I think parse error is closest to what we want from the standard errors defined in jsonrpc spec -> https://www.jsonrpc.org/specification#error_object

**Review by magik6k — APPROVED** [review]

**Comment by Kubuxu** [comment]
Q: Why do we read it into a buffer instead of decoding it directly from the request reader?

**Comment by schomatis** [comment]
> Q: Why do we read it into a buffer instead of decoding it directly from the request reader?

To be able to tell if the client sent more than the limit and report it back as an explicit error instead of just silently truncating it with the limited reader and report a parsing error. Open to other alternatives on how to implement it, though, see the `FIXME`s.

File: handler.go (hunk 1 above)
Comment (jimmylee) [reply handler.go:117]: @magik6k are there any cases where 100MB+ would happen on requests?

Comment (schomatis) [reply handler.go:117]: I'm also fine with this being a configurable option (with a sane default like 100 MiB) for the consumer to adjust as necessary. Not sure what would be the most appropriate way to implement that.

**Comment by Kubuxu** [comment]
Ahh, ignore my question, we were using raw decoder previously.

File: handler.go (hunk 1 above)
Comment (schomatis) [reply handler.go]: Added it as a server configuration, will make Lotus consume it as a node options during dependency update.

**Review by magik6k — APPROVED** [review]

**Review by Kubuxu — APPROVED** [review]

**Pull request merged**